### DISCUSSION TITLE
How to return to an initial view state after user manipulates map?

### DISCUSSION THREAD

**Discussion by ghost** [visgl/deck.gl#7023] [Q&A]
Using DeckGL with React.

I have a map that the user can manipulate using their mouse. I pass a certain `initialViewState` to DeckGL, and I have a button that intends to reset the map to that initialViewState. But this button won't work unless I pass different params to it. Examples below.

Example 1: `resetMap()` doesn't work.

```
const initialViewState = { ... }
const [viewState, setViewState] = useState(initialViewState)

const resetMap = () => setViewState({ ...initialViewState})

<DeckGL
  initialViewState={viewState}
  controller={true}
  layers={...}
>
  <Map />
</DeckGL>
```

Example 2: `resetMap()` works.

```
const initialViewState = { ... }
const [viewState, setViewState] = useState(initialViewState)

// enforce a different latitude
const resetMap = () => setViewState({ ...initialViewState, latitude: initialViewState.latitude + (Math.random() * 0.001) })

<DeckGL
  initialViewState={viewState}
  controller={true}
  layers={...}
>
  <Map />
</DeckGL>
```

Is this just a React issue whereby `viewState`, the state variable, won't change when I use a setter with a similar object? Or is there a more elegant way to implement a simple map reset with DeckGL?

**Comment by mz8i** [comment]
As far as I know, Deck reads `initialViewState`  only once. To update the view state externally, you need to use `viewState` and `onViewStateChange` instead.

Like so:
```js
const initialViewState = { ... }
const [viewState, setViewState] = useState(initialViewState)

// enforce a different latitude
const resetMap = () => setViewState({ ...initialViewState, latitude: initialViewState.latitude + (Math.random() * 0.001) })

<DeckGL
  viewState={viewState}
  onViewStateChange={({viewState}) => setViewState(viewState)}
  controller={true}
  layers={...}
>
  <Map />
</DeckGL>
```

Note there's currently one problematic case - Deck only sanitizes the `viewState` value during initialization (for example to check if the view box is outside valid bounds). It then modifies the view state to fit in the valid bounds and will call `onViewStateChange` with the updated value. But later on, `viewState` is not validated anymore.  Which means that if your initial view state contains values that would get normalized during initialization, resetting the view state to the initial value will cause a shift between the basemap and the deck.gl layers. This issue is tracked here: https://github.com/visgl/deck.gl/issues/6206
